### PR TITLE
Web3signer: Bellatrix Block

### PR DESCRIPTION
**Pull request by james-prysm**
**What type of PR is this?**

Feature

related to #10053

**What does this PR do? Why is it needed?**
Introduce Bellatrix block to web3signer

- add Prysm validatorpb.SignRequest_BlockV3 and validatorpb.SignRequest_BlindedBlockV3 mapping to new web3signer type
- new unit tests for updated request
- update e2e web3signer version
- remote condition to stop web3signer before bellatrix

### CODE REVIEW

File: validator/keymanager/remote-web3signer/keymanager.go [hunk 1]
```diff
@@ -153,22 +153,35 @@ func getSignRequestJson(ctx context.Context, validator *validator.Validate, requ
...
-	// TODO(#10053): Need to add support for merge blocks.
-
-	/*
-		case *validatorpb.SignRequest_BlockV3:
-		return "BLOCK_V3", nil
-	*/
-
+		return json.Marshal(blockv2AltairSignRequest)
+	case *validatorpb.SignRequest_BlockV3:
+		blockv2BellatrixSignRequest, err := v1.GetBlockV2BellatrixSignRequest(request, genesisValidatorsRoot)
```
Comment (rkapka) [inline validator/keymanager/remote-web3signer/keymanager.go:166]: `v1` calls a `V2` function when we have a `V3` case. This is very confusing.

File: validator/keymanager/remote-web3signer/v1/requests.go [hunk 2]
```diff
@@ -263,3 +271,65 @@ func GetSyncCommitteeContributionAndProofSignRequest(request *validatorpb.SignRe
...
+
+// GetBlockV2BellatrixSignRequest maps the request for signing type BLOCK_V2 BELLATRIX.
+func GetBlockV2BellatrixSignRequest(request *validatorpb.SignRequest, genesisValidatorsRoot []byte) (*BlockV2BellatrixSignRequest, error) {
+	if request == nil {
+		return nil, errors.New("nil sign request provided")
+	}
+	var b interfaces.BeaconBlock
+	switch request.Object.(type) {
+	case *validatorpb.SignRequest_BlindedBlockV3:
+		BlindedBlockV3, ok := request.Object.(*validatorpb.SignRequest_BlindedBlockV3)
```
Comment (rkapka) [inline validator/keymanager/remote-web3signer/v1/requests.go:283]: Should be `blindedBlockV3`

File: validator/keymanager/remote-web3signer/v1/requests.go [hunk 3]
```diff
@@ -263,3 +271,65 @@ func GetSyncCommitteeContributionAndProofSignRequest(request *validatorpb.SignRe
...
+		if BlindedBlockV3 == nil {
+			return nil, errors.New("invalid sign request: BlindedBlockV3 is nil")
+		}
+		beaconBlock, err := wrapper.WrappedBeaconBlock(BlindedBlockV3.BlindedBlockV3)
+		if err != nil {
+			return nil, err
+		}
+		b = beaconBlock
+	case *validatorpb.SignRequest_BlockV3:
+		BlockV2Bellatrix, ok := request.Object.(*validatorpb.SignRequest_BlockV3)
```
Comment (rkapka) [inline validator/keymanager/remote-web3signer/v1/requests.go:296]: Should be `blockV2Bellatrix` (and actually it's `V3` - again, this versioning is confusing, why not simply use `bellatrix`?)

File: validator/keymanager/remote-web3signer/v1/requests.go [hunk 4]
```diff
@@ -263,3 +271,65 @@ func GetSyncCommitteeContributionAndProofSignRequest(request *validatorpb.SignRe
...
+	}
+	var b interfaces.BeaconBlock
+	switch request.Object.(type) {
+	case *validatorpb.SignRequest_BlindedBlockV3:
+		BlindedBlockV3, ok := request.Object.(*validatorpb.SignRequest_BlindedBlockV3)
+		if !ok {
+			return nil, errors.New("failed to cast request object to blinded block v3")
+		}
+		if BlindedBlockV3 == nil {
+			return nil, errors.New("invalid sign request: BlindedBlockV3 is nil")
```
Comment (rkapka) [inline validator/keymanager/remote-web3signer/v1/requests.go:288]: Colons separate different errors, using them within errors may cause unwanted parsing. You can replace the colon with a `-` perhaps?

File: validator/keymanager/remote-web3signer/v1/requests.go [hunk 5]
```diff
@@ -263,3 +271,65 @@ func GetSyncCommitteeContributionAndProofSignRequest(request *validatorpb.SignRe
...
+		}
+		b = beaconBlock
+	case *validatorpb.SignRequest_BlockV3:
+		BlockV2Bellatrix, ok := request.Object.(*validatorpb.SignRequest_BlockV3)
+		if !ok {
+			return nil, errors.New("failed to cast request object to block v2 bellatrix")
+		}
+
+		if BlockV2Bellatrix == nil {
+			return nil, errors.New("GetBlockV2BellatrixSignRequest: invalid sign request: BlockV2Bellatrix is nil")
```
Comment (rkapka) [inline validator/keymanager/remote-web3signer/v1/requests.go:302]: same comment about colons in error messages

File: validator/keymanager/remote-web3signer/v1/requests.go [hunk 6]
```diff
@@ -263,3 +271,65 @@ func GetSyncCommitteeContributionAndProofSignRequest(request *validatorpb.SignRe
...
+		}
+		beaconBlock, err := wrapper.WrappedBeaconBlock(BlockV2Bellatrix.BlockV3)
+		if err != nil {
+			return nil, err
+		}
+		b = beaconBlock
+	default:
+		return nil, errors.New("invalid sign request: invalid object type")
```
Comment (rkapka) [inline validator/keymanager/remote-web3signer/v1/requests.go:310]: same comment about colons in error messages

**Review by rkapka — CHANGES_REQUESTED** [review]
Can we simplify the versioning?

File: testing/endtoend/endtoend_test.go [hunk 7]
```diff
@@ -367,14 +367,20 @@ func (r *testRunner) runEvaluators(conns []*grpc.ClientConn, tickingStartTime ti
 // testDepositsAndTx runs tests when config.TestDeposits is enabled.
 func (r *testRunner) testDepositsAndTx(ctx context.Context, g *errgroup.Group,
 	keystorePath string, requiredNodes []e2etypes.ComponentRunner) {
+
+	/// using a shallow copy so I don't mess up something else by changing web3signer setting
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply testing/endtoend/endtoend_test.go]: No I probably should be dereferencing like shallowCopyConfig := *r.config then use a pointer for shallowcopyConfig

File: validator/keymanager/remote-web3signer/v1/requests.go (hunk 6 above)
Comment (james-prysm) [reply validator/keymanager/remote-web3signer/v1/requests.go:310]: good catches

File: testing/endtoend/endtoend_test.go [hunk 8]
```diff
@@ -367,14 +367,19 @@ func (r *testRunner) runEvaluators(conns []*grpc.ClientConn, tickingStartTime ti
...
```
Comment (rauljordan) [inline testing/endtoend/endtoend_test.go:371]:
```suggestion
```

File: validator/keymanager/remote-web3signer/metrics.go [hunk 9]
```diff
@@ -34,6 +34,14 @@ var (
 		Name: "remote_web3signer_block_v2_sign_requests_total",
 		Help: "Total number of block v2 sign requests",
 	})
+	blockV3SignRequestsTotal = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "remote_web3signer_block_v3_sign_requests_total",
```
[the remark this answers is not in the document]
Comment (rauljordan) [reply validator/keymanager/remote-web3signer/metrics.go:38]: I would argue we should stick to the name of the endpoint of the web3signer for this metric in particular, as they use v2 and v3

**Review by rauljordan — APPROVED** [review]

File: validator/keymanager/remote-web3signer/v1/requests.go [hunk 10]
```diff
@@ -263,3 +271,66 @@ func GetSyncCommitteeContributionAndProofSignRequest(request *validatorpb.SignRe
...
+		}
+		b = beaconBlock
+	case *validatorpb.SignRequest_BlockV3:
+		blockV3Bellatrix, ok := request.Object.(*validatorpb.SignRequest_BlockV3)
+		if !ok {
+			return nil, errors.New("failed to cast request object to block v3 bellatrix")
+		}
+
+		if blockV3Bellatrix == nil {
+			return nil, errors.New("GetBlockV2BellatrixSignRequest - invalid sign request: blockV3Bellatrix is nil")
```
Comment (terencechain) [inline validator/keymanager/remote-web3signer/v1/requests.go:303]:
```suggestion
			return nil, errors.New("invalid sign request: blockV3Bellatrix is nil")
```

**Review by rauljordan — DISMISSED** [review]

**Review by nisdas — APPROVED** [review]

**Pull request merged**